### DISCUSSION TITLE
Integration in Unity builds / Single translation unit projects

### DISCUSSION THREAD

**Discussion by jsancheznet** [ocornut/imgui#4646] [New Users Build/Link/Run issues ONLY!]
Hi, i'm failing to integrate Dear ImGui in my project which is just one single .cpp file. I'm using SDL backend  + OpenGL 3 + glad opengl loader.

TLDR: What is the proper way to integrate Dear ImGUI in projects that use unity builds/single translation unit compilation?

The following is the full story.

I started by succesfully compiling and executing example_sdl_opengl3 by rearranging this:
```
#include "imgui.h"
#include "imgui_impl_sdl.h"
#include "imgui_impl_opengl3.h"
```
to this:
```
#include "imgui.cpp"
#include "imgui_tables.cpp"
#include "imgui_widgets.cpp"
#include "imgui_draw.cpp"
#include "imgui_demo.cpp"
#include "imgui_impl_sdl.cpp"
#include "imgui_impl_opengl3.cpp"
``` 

So i naturally tried the same on my project, just to have dear imgui clash with glad.
```
#include "imgui.cpp"
#include "imgui_tables.cpp"
#include "imgui_widgets.cpp"
#include "imgui_draw.cpp"
#include "imgui_demo.cpp"
#include "imgui_impl_opengl3.cpp"
#include "imgui_impl_sdl.cpp"

#include "glad.c"

```
MSVC spits out this error:
```
glad\include\glad/glad.h(27): fatal error C1083: Cannot open include file: '#error OpenGL header already included, remove this include, glad already provides it': No such file or directory
```

So i thought about including glad before Dear ImGui:
```
#include "glad.c"

#include "imgui.cpp"
#include "imgui_tables.cpp"
#include "imgui_widgets.cpp"
#include "imgui_draw.cpp"
#include "imgui_demo.cpp"
#include "imgui_impl_opengl3.cpp"
#include "imgui_impl_sdl.cpp"
```
And then i get big list of warnings for macro redefinitions for each opengl function Dear ImGui loaded plus a few errors.
```
dearimgui\imgui_impl_opengl3_loader.h(551): warning C4005: 'glViewport': macro redefinition
c:\Users\jsanc\Dropbox\Projects\Toasty\libraries\glad\include\glad/glad.h(1100): note: see previous definition of 'glViewport'
c:\Users\jsanc\Dropbox\Projects\Toasty\libraries\dearimgui\imgui_impl_opengl3_loader.h(589): error C2556: 'GL3WglProc get_proc(const char *)': overloaded function differs only by return type from 'void *get_proc(const char *)'
c:\Users\jsanc\Dropbox\Projects\Toasty\glad.c(128): note: see declaration of 'get_proc'
c:\Users\jsanc\Dropbox\Projects\Toasty\libraries\dearimgui\imgui_impl_opengl3_loader.h(588): error C2373: 'get_proc': redefinition; different type modifiers
c:\Users\jsanc\Dropbox\Projects\Toasty\glad.c(128): note: see declaration of 'get_proc'
c:\Users\jsanc\Dropbox\Projects\Toasty\libraries\dearimgui\imgui_impl_opengl3_loader.h(682): error C2065: 'get_proc': undeclared identifier
```

If i remove glad from my project, gladLoadGL is not defined and i cant call OpenGL functions.

What is the proper way to integrate Dear ImGUI in projects that use unity builds/single translation unit compilation?

thank you for your time.

**Comment by PathogenDavid** [comment]
As of 1.84 Dear ImGui's OpenGL 3 backend uses its own minimal OpenGL loader, you can read more details about this here: https://github.com/ocornut/imgui/issues/4445

In your case, the Dear ImGui loader and Glad are conflicting since you're building them in the same translation unit.

I believe the easiest workaround in your case would be to define `IMGUI_IMPL_OPENGL_LOADER_CUSTOM` so that the backend picks up Glad instead.

**Comment by jsancheznet** [comment]
@PathogenDavid Defining ```IMGUI_IMPL_OPENGL_LOADER_CUSTOM``` worked like a charm.

Thank you!
